### PR TITLE
Allow multiple process per device

### PR DESCRIPTION
**Pull request by cifkao**
Running multiple replicas of a model on each device can speed up inference. However, `accelerate` currently does not seem to support this. Setting `num_processes` to a value larger than the number of available devices results in an error (for CUDA: "invalid device ordinal").

# What does this PR do?

Makes sure that the device index is always valid by taking the remainder upon division by `device_count()`, hence allowing multiple processes to run on each device by setting `num_processes` to a multiple of `device_count()`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

@muellerzr @BenjaminBossan @SunMarc

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks for the PR. Do you have an example code snippet that produces the error you mentioned?

**Comment by cifkao** [comment]
@BenjaminBossan
I adapted the [stable diffusion example](https://huggingface.co/docs/accelerate/en/usage_guides/distributed_inference#sending-chunks-of-a-batch-automatically-to-each-loaded-model) from the docs:
```python
from accelerate import PartialState
from diffusers import DiffusionPipeline
import torch

pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
distributed_state = PartialState()
pipe.to(distributed_state.device)

# Assume two processes
with distributed_state.split_between_processes(["a dog", "a cat", "a snail", "a hedgehog"]) as prompt:
    print(f"Running {prompt} on process {distributed_state.process_index}, device {distributed_state.device}")
    result = pipe(prompt).images
    result[0].save(f"{prompt[0]}.png")
```

Command:
```bash
CUDA_VISIBLE_DEVICES=0,1 accelerate launch --num-processes 4 distributed_inference.py
```

<details>
<summary>Error:</summary>

```
[rank3]: Traceback (most recent call last):
[rank3]:   File "/home/ondrej/proj/accelerate/distributed_inference.py", line 6, in <module>
[rank3]:     distributed_state = PartialState()
[rank3]:                         ^^^^^^^^^^^^^^
[rank3]:   File "/home/ondrej/proj/accelerate/src/accelerate/state.py", line 280, in __init__
[rank3]:     self.set_device()
[rank3]:   File "/home/ondrej/proj/accelerate/src/accelerate/state.py", line 790, in set_device
[rank3]:     torch.cuda.set_device(self.device)
[rank3]:   File "/home/ondrej/mambaforge/envs/accelerate/lib/python3.11/site-packages/torch/cuda/__init__.py", line 399, in set_device
[rank3]:     torch._C._cuda_setDevice(device)
[rank3]: RuntimeError: CUDA error: invalid device ordinal
[rank3]: CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
[rank3]: For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
[rank3]: Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

[rank2]: Traceback (most recent call last):
[rank2]:   File "/home/ondrej/proj/accelerate/distributed_inference.py", line 6, in <module>
[rank2]:     distributed_state = PartialState()
[rank2]:                         ^^^^^^^^^^^^^^
[rank2]:   File "/home/ondrej/proj/accelerate/src/accelerate/state.py", line 280, in __init__
[rank2]:     self.set_device()
[rank2]:   File "/home/ondrej/proj/accelerate/src/accelerate/state.py", line 790, in set_device
[rank2]:     torch.cuda.set_device(self.device)
[rank2]:   File "/home/ondrej/mambaforge/envs/accelerate/lib/python3.11/site-packages/torch/cuda/__init__.py", line 399, in set_device
[rank2]:     torch._C._cuda_setDevice(device)
[rank2]: RuntimeError: CUDA error: invalid device ordinal
[rank2]: CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
[rank2]: For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
[rank2]: Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

W0706 11:17:44.554000 139756795762496 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 4070667 closing signal SIGTERM
W0706 11:17:44.555000 139756795762496 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 4070670 closing signal SIGTERM
W0706 11:17:44.555000 139756795762496 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 4070671 closing signal SIGTERM
E0706 11:17:45.083000 139756795762496 torch/distributed/elastic/multiprocessing/api.py:826] failed (exitcode: 1) local_rank: 3 (pid: 4070672) of binary: /home/ondrej/mambaforge/env
s/accelerate/bin/python3.11
```
</details>

With this patch, it outputs:
```
Running ['a dog'] on process 0, device cuda:0
Running ['a snail'] on process 2, device cuda:0
Running ['a hedgehog'] on process 3, device cuda:1
Running ['a cat'] on process 1, device cuda:1
```

**Comment by BenjaminBossan** [comment]
Thanks for providing the example. I wonder if it could be adapted as a unit test on a multi GPU runner, assuming we want to proceed with the PR as is. I'll let @muellerzr decide this.

> assert device in ("cpu", "cuda", "mlu", "npu", "xpu")

I think this line can safely be removed, as it's covered by the `ValueError` a few lines above.

File: src/accelerate/state.py
Comment (muellerzr) [inline src/accelerate/state.py:781]: We generally don't do hard asserts like this here, and as Benjamin pointed out we should be covered for this.
```suggestion
```

**Review by muellerzr — COMMENTED** [review]
Thanks for the improvement!

If you can add a test perhaps in `src/accelerate/test_utils/scripts/test_ops.py` most likely, or in `external_deps/test_metrics`?

**Comment by muellerzr** [comment]
That makes me a bit hesitant to push this in, since generally users will likely want to do `.gather()` on all the outputs after inference. How are you collecting the outputs from the model to a centralized location?

**Comment by cifkao** [comment]
I just save the files to a directory as in the example in the docs. I agree it would be better to have support for `gather`...

Related discussions I found:
- https://github.com/pytorch/ignite/issues/2139
- https://discuss.pytorch.org/t/reusing-a-gpu-by-multiple-distributed-processes/137394

**Review by muellerzr — APPROVED** [review]
That makes sense. As long as our main tests don't break, and your use-case works, I don't see an issue to why we can't merge this in since that's what we really care about.

Do you have two GPUs to run `RUN_SLOW=1 pytest -sv tests/` with?

**Comment by cifkao** [comment]
I get the same set of failed tests for `main` and this branch:
```
FAILED tests/deepspeed/test_deepspeed.py::DeepSpeedConfigIntegration::test_autofill_comm_buffers_dsconfig_config_with_hidden_size - deepspeed.ops.op_builder.builder.CUDAMismatc
hException: >- DeepSpeed Op Builder: Installed CUDA version 10.1 does not match the version torch was compiled with 12.1, unabl...
FAILED tests/deepspeed/test_deepspeed.py::DeepSpeedConfigIntegration::test_autofill_comm_buffers_dsconfig_config_with_hidden_sizes - deepspeed.ops.op_builder.builder.CUDAMismat
chException: >- DeepSpeed Op Builder: Installed CUDA version 10.1 does not match the version torch was compiled with 12.1, unabl...
FAILED tests/deepspeed/test_deepspeed.py::DeepSpeedConfigIntegration::test_prepare_deepspeed_custom_optimizer_custom_scheduler - deepspeed.ops.op_builder.builder.CUDAMismatchEx
ception: >- DeepSpeed Op Builder: Installed CUDA version 10.1 does not match the version torch was compiled with 12.1, unabl...
FAILED tests/deepspeed/test_deepspeed.py::DeepSpeedConfigIntegration::test_prepare_deepspeed_custom_optimizer_deepspeed_scheduler - deepspeed.ops.op_builder.builder.CUDAMismatc
hException: >- DeepSpeed Op Builder: Installed CUDA version 10.1 does not match the version torch was compiled with 12.1, unabl...
FAILED tests/test_accelerator.py::AcceleratorTester::test_get_state_dict_from_offload_use_pytorch - ValueError: cannot mmap an empty file
FAILED tests/test_accelerator.py::AcceleratorTester::test_get_state_dict_from_offload_use_safetensors - ValueError: cannot mmap an empty file
FAILED tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_pytorch - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_safetensors - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_accelerator.py::AcceleratorTester::test_save_sharded_model_use_pytorch - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_accelerator.py::AcceleratorTester::test_save_sharded_model_use_safetensors - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_cpu_offload - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_cpu_offload_with_unused_submodules - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_disk_offload - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_disk_offload_with_unused_submodules - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_and_remove_hook - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_copy - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_force_hooks - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_multi_devices - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_with_non_persistent_buffers - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_with_unused_submodules - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_dispatch_model_with_unused_submodules_multi_device - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_init_empty_weights - NotImplementedError: Cannot copy out of meta tensor; no data!
FAILED tests/test_big_modeling.py::BigModelingTester::test_load_checkpoint_and_dispatch - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_load_checkpoint_and_dispatch_multi_device - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_load_checkpoint_and_dispatch_multi_device_with_unused_submodules - RuntimeError: weight should contain 5 elements not
 0
FAILED tests/test_big_modeling.py::BigModelingTester::test_load_checkpoint_and_dispatch_with_unused_submodules - RuntimeError: weight should contain 5 elements not 0
FAILED tests/test_cli.py::AccelerateLauncherTester::test_accelerate_test - RuntimeError: 'accelerate test' failed with returncode 1                                 [8592/45802]
FAILED tests/test_cli.py::ModelEstimatorTester::test_gated - AssertionError: GatedRepoError not raised : Repo for model `meta-llama/Llama-2-7b-hf` is gated
FAILED tests/test_examples.py::FeatureExamplesTests::test_checkpointing_by_epoch - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_fi
le=/tmp/tmpyq6p0qhv/default_config.yml examples/by_feature/checkpointing.py --...
FAILED tests/test_examples.py::FeatureExamplesTests::test_checkpointing_by_steps - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_fi
le=/tmp/tmpyq6p0qhv/default_config.yml examples/by_feature/checkpointing.py --...
FAILED tests/test_examples.py::FeatureExamplesTests::test_cross_validation - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=/tm
p/tmpyq6p0qhv/default_config.yml examples/by_feature/cross_validation.py...
FAILED tests/test_examples.py::FeatureExamplesTests::test_ddp_comm_hook - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=/tmp/t
mpyq6p0qhv/default_config.yml examples/by_feature/ddp_comm_hook.py --...
FAILED tests/test_examples.py::FeatureExamplesTests::test_distributed_inference_examples_phi2 - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launc
h --config_file=/tmp/tmpyq6p0qhv/default_config.yml examples/inference/distributed/phi2.py`...
FAILED tests/test_examples.py::FeatureExamplesTests::test_distributed_inference_examples_stable_diffusion - accelerate.test_utils.testing.SubprocessCallException: Command `acce
lerate launch --config_file=/tmp/tmpyq6p0qhv/default_config.yml examples/inference/distributed/stable_d...
FAILED tests/test_examples.py::FeatureExamplesTests::test_early_stopping - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=/tmp/
tmpyq6p0qhv/default_config.yml examples/by_feature/early_stopping.py` ...
FAILED tests/test_examples.py::FeatureExamplesTests::test_gradient_accumulation - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_fil
e=/tmp/tmpyq6p0qhv/default_config.yml examples/by_feature/gradient_accumulati...
FAILED tests/test_examples.py::FeatureExamplesTests::test_load_states_by_epoch - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file
=/tmp/tmpyq6p0qhv/default_config.yml examples/by_feature/checkpointing.py --...
FAILED tests/test_examples.py::FeatureExamplesTests::test_load_states_by_steps - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file
=/tmp/tmpyq6p0qhv/default_config.yml examples/by_feature/checkpointing.py --...
FAILED tests/test_examples.py::FeatureExamplesTests::test_local_sgd - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=/tmp/tmpyq
6p0qhv/default_config.yml examples/by_feature/local_sgd.py` faile...
FAILED tests/test_examples.py::FeatureExamplesTests::test_multi_process_metrics - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_fil
e=/tmp/tmpyq6p0qhv/default_config.yml examples/by_feature/multi_process_metri...
FAILED tests/test_examples.py::FeatureExamplesTests::test_pippy_examples_bert - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=
/tmp/tmpyq6p0qhv/default_config.yml examples/inference/pippy/bert.py` faile...
FAILED tests/test_examples.py::FeatureExamplesTests::test_pippy_examples_gpt2 - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=
/tmp/tmpyq6p0qhv/default_config.yml examples/inference/pippy/gpt2.py` faile...
FAILED tests/test_examples.py::FeatureExamplesTests::test_pippy_examples_t5 - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=/t
mp/tmpyq6p0qhv/default_config.yml examples/inference/pippy/t5.py` failed ...
FAILED tests/test_examples.py::FeatureExamplesTests::test_profiler - accelerate.test_utils.testing.SubprocessCallException: Command `accelerate launch --config_file=/tmp/tmpyq6
p0qhv/default_config.yml examples/by_feature/profiler.py` failed...
FAILED tests/test_grad_sync.py::SyncScheduler::test_gradient_sync_gpu_multi - RuntimeError: 'accelerate launch --num_processes=2 --monitor_interval=0.1 /home/ondrej/proj/accele
rate/src/accelerate/test_utils/scripts/test_sync.py' failed with returnco...
FAILED tests/test_hooks.py::HooksModelTester::test_add_remove_hook_fx_graph_module - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_hooks.py::HooksModelTester::test_align_devices_as_cpu_offload - AssertionError: assert device(type='cuda', index=0) == device(type='cpu')
FAILED tests/test_hooks.py::HooksModelTester::test_align_devices_as_model_parallelism - AssertionError: assert device(type='cuda', index=0) == device(type='cpu')
FAILED tests/test_hooks.py::HooksModelTester::test_attach_align_device_hook_as_cpu_offload - AssertionError: assert device(type='cuda', index=0) == device(type='cpu')
FAILED tests/test_hooks.py::HooksModelTester::test_attach_align_device_hook_as_cpu_offload_with_weight_map - AssertionError: assert device(type='cuda', index=0) == device(type=
'cpu')
FAILED tests/test_hooks.py::HooksModelTester::test_no_grad_in_hook - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_hooks.py::HooksModelTester::test_post_forward_hook_is_executed - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_hooks.py::HooksModelTester::test_pre_forward_hook_is_executed - RuntimeError: weight should contain 4 elements not 0
FAILED tests/test_kwargs_handlers.py::KwargsHandlerTester::test_ddp_comm_hook - RuntimeError: 'accelerate launch --num_processes=2 --monitor_interval=0.1 /home/ondrej/proj/acce
lerate/src/accelerate/test_utils/scripts/test_ddp_comm_hook.py' failed with...
FAILED tests/test_kwargs_handlers.py::KwargsHandlerTester::test_ddp_kwargs - RuntimeError: 'accelerate launch --num_processes=2 --monitor_interval=0.1 /home/ondrej/proj/acceler
ate/tests/test_kwargs_handlers.py' failed with returncode 1
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_compute_module_sizes - AssertionError: assert defaultdict(<..._tracked': 8}) == {'': 236, 'ba...cked': 8, ...}
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_get_balanced_memory - AssertionError: assert {0: 200, 1: 200} == {0: 187, 1: 200}
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_infer_auto_device_map_with_buffer_check_and_multi_devices - assert 1 == 0
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_infer_auto_device_map_with_tied_weights - AssertionError: assert OrderedDict([...tchnorm', 1)]) == {'layer1': 0,.
..ear1': 1, ...}
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_load_checkpoint_in_model_disk_offload - ValueError: cannot mmap an empty file
FAILED tests/test_offload.py::OffloadTester::test_offload_state_dict - ValueError: cannot mmap an empty file
FAILED tests/test_offload.py::OffloadTester::test_offload_weights_loader - ValueError: cannot mmap an empty file
==================================================== 63 failed, 235 passed, 40 skipped, 519 warnings in 1153.61s (0:19:13) =====================================================

**Comment by muellerzr** [comment]
Great sounds good! Thanks for the improvement!

**Pull request merged**
